lib/iris/experimental/ugrid.py

```diff
@@ -1680,102 +1679,316 @@ def remove(
-#         pass
-#
-#     def __eq__(self, other):
-#         # Full equality could be MASSIVE, so we want to avoid that.
-#         # Ideally we want a mesh signature from LFRic for comparison, although this would
-#         # limit Iris' relevance outside MO.
-#         # TL;DR: unknown quantity.
-#         raise NotImplemented
-#
-#     def __ne__(self, other):
-#         # See __eq__
-#         raise NotImplemented
-#
-#
-# class _Mesh2DConnectivityManager(_Mesh1DConnectivityManager):
-#     REQUIRED = (
-#         "face_node",
-#     )
-#     OPTIONAL = (
-#         "edge_node",
-#         "face_edge",
-#         "face_face",
-#         "edge_face",
-#         "boundary_node",
-#     )
-#     def __init__(self, face_node, edge_node=None, face_edge=None, face_face=None, edge_face=None, boundary_node=None):
-#         # required
-#         self.face_node = face_node
-#         self._members = [self.face_node]
-#
-#         # optionally required
-#         self.edge_node = edge_node
-#         # optional
-#         self.face_edge = face_edge
-#         self.face_face = face_face
-#         self.edge_face = edge_face
-#         self.boundary_node = boundary_node
-#
-#         # edge_node could be None here. are we okay with this pattern?
-#         super().__init__(edge_node)
-#
-#         # does order matter?
-#         self._members.extend([member for member in self.OPTIONAL if member is not None and member != "edge_node"])
+class _MeshConnectivityManagerMixin(ABC):
+    REQUIRED = ()
+    OPTIONAL = ()
+    NDIM = NotImplemented
+
+    @abstractmethod
+    def __init__(self, *connectivities):
+        cf_roles = [c.cf_role for c in connectivities]
+        for requisite in self.REQUIRED:
+            if requisite not in cf_roles:
+                message = f"{self.NDIM}D meshes require a {requisite}."
+                raise ValueError(message)
+
+        self.ALL = self.REQUIRED + self.OPTIONAL
+        self._members = {member: None for member in self.ALL}
+        self.add(*connectivities)
+
+    def __eq__(self, other):
+        # TBD
+        return NotImplemented
+
+    def __getstate__(self):
+        # TBD
+        pass
+
+    def __iter__(self):
+        for item in self._members.items():
+            yield item
+
+    def __ne__(self, other):
+        # TBD
+        return NotImplemented
+
+    def __repr__(self):
+        args = [
+            f"{member}={connectivity!r}"
+            for member, connectivity in self
+            if connectivity is not None
+        ]
+        return f"{self.__class__.__name__}({', '.join(args)})"
+
+    def __setstate__(self, state):
+        # TBD
+        pass
+
+    def __str__(self):
+        args = [
+            f"{member}=True"
+            for member, connectivity in self
+            if connectivity is not None
+        ]
+        return f"{self.__class__.__name__}({', '.join(args)})"
+
+    @property
+    @abstractmethod
+    def all_members(self):
+        return NotImplemented
```

Comment:
@trexfeathers As mentioned above, this is an abstract base class or a mixin.

If you want this to be a mixin, then there is no need for the definition of the `all_members` method.

Reply:
f5be558813e7c30ab24ac903d99c62dc2817aa67